pkg/protoutil/crypt.go

```diff
@@ -0,0 +1,169 @@
+package protoutil
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/pomerium/pomerium/pkg/cryptutil"
+	cryptpb "github.com/pomerium/pomerium/pkg/grpc/crypt"
+)
+
+// An Encryptor encrypts protobuf messages using a key encryption key and periodically rotated
+// generated data encryption keys.
+type Encryptor struct {
+	kek         *cryptutil.PublicKeyEncryptionKey
+	rotateEvery time.Duration
+
+	sync.RWMutex
+	nextRotate   time.Time
+	dek          *cryptutil.DataEncryptionKey
+	encryptedDEK []byte
+}
+
+// NewEncryptor returns a new protobuf Encryptor.
+func NewEncryptor(kek *cryptutil.PublicKeyEncryptionKey) *Encryptor {
+	return &Encryptor{
+		kek:         kek,
+		rotateEvery: time.Hour,
+	}
+}
+
+func (enc *Encryptor) getDataEncryptionKey() (*cryptutil.DataEncryptionKey, []byte, error) {
+	// double-checked locking
+	// first time we do a read only lookup
+	enc.RLock()
+	dek, encryptedDEK, err := enc.getDataEncryptionKeyLocked(true)
+	enc.RUnlock()
+	if err != nil {
+		return nil, nil, err
+	} else if dek != nil {
+		return dek, encryptedDEK, nil
+	}
+
+	// second time we do a read/write lookup
+	enc.Lock()
+	dek, encryptedDEK, err = enc.getDataEncryptionKeyLocked(false)
+	enc.Unlock()
+	return dek, encryptedDEK, err
+}
+
+func (enc *Encryptor) getDataEncryptionKeyLocked(readOnly bool) (*cryptutil.DataEncryptionKey, []byte, error) {
+	needsNewKey := enc.dek == nil || time.Now().After(enc.nextRotate)
+	if !needsNewKey {
+		return enc.dek, enc.encryptedDEK, nil
+	}
+
+	if readOnly {
+		return nil, nil, nil
+	}
+
+	// generate a new data encryption key
+	dek, err := cryptutil.GenerateDataEncryptionKey()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// seal the data encryption key using the key encryption key
+	encryptedDEK, err := enc.kek.EncryptDataEncryptionKey(dek)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	enc.dek = dek
+	enc.encryptedDEK = encryptedDEK
+	enc.nextRotate = time.Now().Add(enc.rotateEvery)
+
+	return enc.dek, enc.encryptedDEK, nil
+}
+
+// Encrypt encrypts a protobuf message.
+func (enc *Encryptor) Encrypt(msg proto.Message) (*cryptpb.SealedMessage, error) {
+	// get the data encryption key
+	dek, encryptedDEK, err := enc.getDataEncryptionKey()
+	if err != nil {
+		return nil, err
+	}
+
+	plaintext, err := protojson.Marshal(msg)
```

Comment:
why serialize protobuf to JSON?

Reply:
The intention here was that if someone were decrypting the payloads through some other mechanism, it'd be easier to read JSON than raw protobuf, since it's self-describing and most log systems seem to have native support for it.